nengo_gui/components/netgraph.py

```diff
@@ -580,24 +583,81 @@ def send_pan_and_zoom(self, client):
         client.write_text(json.dumps(dict(type='pan', pan=pan)))
         client.write_text(json.dumps(dict(type='zoom', zoom=zoom)))
 
-    def create_connection(self, client, conn, parent):
-        uid = self.page.get_uid(conn)
-        if uid in self.uids:
-            return
+    @staticmethod
+    def connection_pre_obj(conn):
+        """
+        Returns the pre-object of the given connection. Tries to make sure that
+        the pre-object is an Ensemble. For example, if the connection originates
+        from the ".neurons" object, the ensemble corresponding to the raw
+        neurons is returned.
+        """
         pre = conn.pre_obj
         if isinstance(pre, nengo.ensemble.Neurons):
             pre = pre.ensemble
+        return pre
+
+    @staticmethod
+    def connection_post_obj(conn):
+        """
+        Returns the post-object relevant to the visualisation of the given
+        connection. If the connection is a learning rule, returns the
+        post-ensemble of that connection. If the connection ends in a
+        ".neurons" object, returns the corresponding ensemble.
+        """
         post = conn.post_obj
         if isinstance(post, nengo.connection.LearningRule):
             post = post.connection.post
             if isinstance(post, nengo.base.ObjView):
                 post = post.obj
         if isinstance(post, nengo.ensemble.Neurons):
             post = post.ensemble
-        pre = self.page.get_uid(pre)
-        post = self.page.get_uid(post)
+        return post
+
+    @staticmethod
+    def connection_kind(conn):
+        """
+        Categorises the given connection into one of three kinds: modulatory,
+        inhibitory, and normal.
+        """
+        if isinstance(conn.post_obj, nengo.connection.LearningRule):
+            return "modulatory"
+        if isinstance(conn.post_obj, nengo.ensemble.Neurons):
+            trafo = conn.transform
+            if trafo.size > 0 and np.all(trafo < 0):
```

Comment:
I wonder if this should rather be `trafo <= 0`?

Reply:
Hm, I don't think so. I wouldn't consider a connection to be inhibitory if the connection has no effect whatsoever.

Reply:
`np.all(trafo <= 0)` allows for a connection that has not effect, but it doesn't imply that the connection has no effect. I would consider a sparse inhibitory matrix still inhibitory.

Reply:
Ok, I see, you are right. I'm going for

```python
np.all(trafo <= 0.0) and not np.all(np.isclose(trafo, 0.0))
```

and add some more test cases.

Reply:
Ok, fixed this, as discussed above.

Reply:
You can use `np.allclose` instead of `np.all(np.isclose(...))`.